src/aktualizr_secondary/aktualizr_secondary.cc

```diff
@@ -74,15 +75,43 @@ std::pair<KeyType, std::string> AktualizrSecondary::getPublicKeyResp() const {
   return std::make_pair(config_.uptane.key_type, keys_.getUptanePublicKey());
 }
 
-Json::Value AktualizrSecondary::getManifestResp() const {
-  LOG_ERROR << "getManifestResp is not implemented yet";
-  return Json::Value();
-}
+Json::Value AktualizrSecondary::getManifestResp() const { return pacman->getManifest(getSerialResp()); }
 
 bool AktualizrSecondary::putMetadataResp(const Uptane::MetaPack &meta_pack) {
-  (void)meta_pack;
-  LOG_ERROR << "putMedatadatResp is not implemented yet";
-  return false;
+  Uptane::TimeStamp now(Uptane::TimeStamp::Now());
+  detected_attack_.clear();
+
+  root_ = Uptane::Root(now, "director", meta_pack.director_root.original(), root_);
+  Uptane::Targets targets(now, "director", meta_pack.director_targets.original(), root_);
+  if (meta_targets_.version() > targets.version()) {
+    detected_attack_ = "Rollback attack detected";
+    return true;
+  }
+  meta_targets_ = targets;
+  std::vector<Uptane::Target>::const_iterator it;
+  bool target_found = false;
+  const Uptane::Target *target;
+  for (it = meta_targets_.targets.begin(); it != meta_targets_.targets.end(); ++it) {
+    if (it->IsForSecondary(getSerialResp())) {
+      if (target_found) {
+        detected_attack_ = "Duplicate entry for this ECU";
+        break;
+      }
+      target_found = true;
+      target = &(*it);
+    }
+  }
+  if (target_found) {
+    if (target->format().empty() || target->format() == "OSTREE") {
+#ifdef BUILD_OSTREE
+      OstreeManager::pull(config_.pacman, keys_, target->sha256Hash());
+#else
+      LOG_ERROR << "Could not pull OSTree target. Aktualizr was built without OSTree support!";
+#endif
+    }
+    pacman->install(*target);
```

Comment:
Installation should only happen in `sendFirmwareResp`. It gets TLS credentials as a payload and without them OSTree will not be able to download the image.

Reply:
Ok